### ISSUE TITLE
Question about "https://github.com/openvinotoolkit/deep-object-reid/tree/multilabel"

### ISSUE THREAD

**Issue by UnityBoy** [open-edge-platform/training_extensions#1609]
Hi,there.

the quesition is about the training method of "https://github.com/openvinotoolkit/deep-object-reid/tree/multilabel" REPO. But i cannot find an issue on that repo.

question:

firsrtly,
it is said we can training by used this code?  but are you sure? " ['<data_root>/train.json', '<data_root>/train.json']"  

the list[0] and list[1] are the same?

python tools/main.py --config-file configs/EfficientNetV2_small_gcn.yml --gpu-num 1 custom_datasets.roots "['<data_root>/train.json', '<data_root>/train.json']" data.save_dir <work_dir> 

second,
I used VOC2007 datasets for traning and it is reported a  bug just after 16/60 epoch ,it is said:"LOG::The training should be stopped due to no improvements for 5 epochs" when i change the code to 
 " ['<data_root>/train.json', '<data_root>/val.json']"

2023-02-01_19:03:34|epoch: [16/60][59/59]       time 0.985 (1.177)      data 0.044 (0.108)      cls acc 0.833 (0.883)    eta 0:50:55     main_model 0.8399 (0.8703)      loss 0.8399 (0.8703)    lr 0.006881
2023-02-01_19:03:34|##### Evaluating test dataset #####
2023-02-01_19:04:01|** Results (main_model) **
2023-02-01_19:04:01|mAP: 94.91%
2023-02-01_19:04:01|P_O: 90.70%
2023-02-01_19:04:01|R_O: 91.38%
2023-02-01_19:04:01|F_O: 91.04%
2023-02-01_19:04:01|mean_P_C: 92.97%
2023-02-01_19:04:01|mean_R_C: 94.03%
2023-02-01_19:04:01|mean_F_C: 92.24%
2023-02-01_19:04:29|** Results (EMA model) **
2023-02-01_19:04:29|mAP: 91.25%
2023-02-01_19:04:29|P_O: 98.92%
2023-02-01_19:04:29|R_O: 35.93%
2023-02-01_19:04:29|F_O: 52.71%
2023-02-01_19:04:29|mean_P_C: 52.56%
2023-02-01_19:04:29|mean_R_C: 39.69%
2023-02-01_19:04:29|mean_F_C: 43.65%
2023-02-01_19:04:29|LOG:: The training should be stopped due to no improvements for 5 epochs
2023-02-01_19:04:29|Elapsed 0:32:41

thanks 

best regards

**Comment by sovrasov** [comment]
Hi @UnityBoy, 
1. Thanks for pointing out this mistake in tutorial, I'll fix it.
2. This message is expected, it's not a bug. It indicates that an early stopping criterion has been fulfilled during the training.

**Comment by UnityBoy** [comment]
> Hi @UnityBoy,
> 
> 1. Thanks for pointing out this mistake in tutorial, I'll fix it.
> 2. This message is expected, it's not a bug. It indicates that an early stopping criterion has been fulfilled during the training.



Yeah, but it was always  early stopping  . Could you show more about how did you training?I just used default training setting .Thx

here are my training setting:

CUDA_VISIBLE_DEVICES="4,6,7" python tools/main.py --config-file configs/Efficien                                                                 tNetV2_small_mld.yml --gpu-num 3 custom_datasets.roots "['datasets/VOC/train.json', 'datasets/VOC/val.json']" dat                                                                 a.save_dir output/voc/v4

datasets also as same as default VOC2007


/home/multilabel/torchreid/__init__.py:14: UserWarning: NNCF was not found. Model optimization options                                                                  will not be available.
  warnings.warn("NNCF was not found. Model optimization options will not be available.")
Begin merging of config files with inheritance
    merging config file configs/EfficientNetV2_small_mld.yml
End merging of config files with inheritance
fatal: not a git repository (or any of the parent directories): .git
fatal: not a git repository (or any of the parent directories): .git
2023-02-01_17:29:08|HEAD is:
2023-02-01_17:29:08|commit SHA is:
2023-02-01_17:29:08|
2023-02-01_17:29:08|Show configuration
_base_:
adam:
  beta1: 0.9
  beta2: 0.999
custom_datasets:
  names: []
  roots: ['datasets/VOC/train.json', 'datasets/VOC/train.json']
  types: ['multilabel_classification', 'multilabel_classification']
data:
  combineall: False
  height: 448
  min_samples_per_id: 1
  norm_mean: [0.5, 0.5, 0.5]
  norm_std: [0.5, 0.5, 0.5]
  num_sampled_packages: 1
  root: ./
  save_dir: output/voc/v3
  split_id: 0
  tb_log_dir:
  transforms:
    augmix:
      cfg_str: augmix-m5-w3
      enable: False
      grey_imgs: False
    center_crop:
      enable: False
      margin: 0
      test_only: False
    coarse_dropout:
      enable: False
      fill_value: 0
      mask_fill_value: None
      max_height: 8
      max_holes: 8
      max_width: 8
      min_height: None
      min_holes: None
      p: 0.5
    color_jitter:
      brightness: 0.2
      contrast: 0.2
      enable: False
      hue: 0.1
      p: 0.5
      saturation: 0.1
    crop_pad:
      enable: False
    cutout:
      cutout_factor: 0.4
      enable: True
      fill_color: random
      p: 0.5
    equalize:
      enable: False
      p: 0.5
    force_gray_scale:
      enable: False
    posterize:
      bits: 1
      enable: False
      p: 0.5
    randaugment:
      enable: True
      p: 1.0
    random_blur:
      enable: False
      k: 5
      p: 0.5
    random_crop:
      align_ar: False
      align_center: False
      enable: False
      margin: 0
      p: 0.5
      scale: 0.9
      static: False
    random_erase:
      enable: False
      fill_color: (125.307, 122.961, 113.8575)
      norm_image: True
      p: 0.5
      rh: 3.3
      rl: 0.3
      sh: 0.4
      sl: 0.2
    random_figures:
      always_single_figure: False
      circle_radiuses: (5, 64)
      enable: False
      figure_prob: 0.5
      figures: ['line', 'rectangle', 'circle']
      p: 0.5
      random_color: True
      thicknesses: (1, 6)
    random_flip:
      enable: True
      p: 0.5
    random_gray_scale:
      enable: False
      p: 0.5
    random_negative:
      enable: False
      p: 0.5
    random_noise:
      enable: False
      grayscale: False
      p: 0.2
      sigma: 0.05
    random_perspective:
      distortion_scale: 0.5
      enable: False
      p: 0.5
    random_rotate:
      angle: (-5, 5)
      enable: False
      p: 0.5
      values: (0,)
    test:
      resize_first: False
      resize_scale: 1.0
  width: 448
  workers: 6
loss:
  am_binary:
    amb_k: 0.7
    amb_t: 1.0
  asl:
    gamma_neg: 0.0
    gamma_pos: 0.0
    p_m: 0.05
  name: am_binary
  softmax:
    augmentations:
      alpha: 1.0
      aug_prob: 1.0
      aug_type:
      fmix:
        decay_power: 3
    compute_s: False
    conf_penalty: 0.0
    label_smooth: 0.0
    m: 0.0
    margin_type: cos
    pr_product: False
    s: 23.0
    symmetric_ce: False
lr_finder:
  enable: False
  epochs_warmup: 1
  max_lr: 0.001
  min_lr: 1e-05
  mode: TPE
  n_trials: 15
  num_epochs: 10
  path_to_savefig: lr_finder.jpg
  smooth_f: 0.01
  step: 1e-05
  stop_after: False
mixing_loss:
  enable: False
  weight: 1.0
model:
  IN_conv1: False
  IN_first: False
  bn_eval: False
  bn_frozen: False
  download_weights: True
  dropout_backbone:
    dist: none
    kernel: 3
    mu: 0.1
    p: 0.0
    sigma: 0.03
    temperature: 0.2
  dropout_cls:
    dist: none
    kernel: 3
    mu: 0.1
    p: 0.1
    sigma: 0.03
    temperature: 0.2
  export_onnx_opset: 13
  feature_dim: 512
  gcn:
    adj_matrix_path:
    hidden_dim_scale: 1.0
    layer_type: gcn
    rho: 0.25
    thau: 0.4
    word_emb_path:
    word_model_path:
  load_weights: efficientnetv2-s-oi.pth
  name: mld_efficientnetv2_s_21k
  pooling_type: avg
  pretrained: True
  resume:
  save_all_chkpts: False
  self_challenging_cfg:
    drop_batch_p: 0.33
    drop_p: 0.33
    enable: False
  transformer:
    dropout: 0.1
    nheads: 4
    num_decoder_layers: 2
    num_encoder_layers: 1
    pre_norm: False
    rm_first_self_attn: True
    rm_self_attn_dec: True
  type: multilabel
mutual_learning:
  aux_configs: []
nncf:
  coeff_decrease_lr_for_nncf: 0.035
  nncf_config_path:
rmsprop:
  alpha: 0.99
sampler:
  train_sampler: RandomSampler
sc_integration:
  epoch_scale: 1.0
  lr_scale: 1.0
sgd:
  dampening: 0.0
  momentum: 0.9
  nesterov: False
test:
  batch_size: 128
  estimate_multilabel_thresholds: False
  eval_freq: 1
  evaluate: False
  save_initial_metric: False
  start_eval: 0
  test_before_train: False
  topk: [1, 5, 10, 20]
train:
  base_lr_mult: 0.1
  base_optim: sgd
  base_scheduler:
  batch_size: 84
  clip_grad: 0.0
  correct_batch_size: False
  cycle_mult: 1.0
  deterministic: True
  early_stopping: True
  ema:
    ema_decay: 0.9997
    enable: True
  first_cycle_steps: 5
  fixbase_epoch: 0
  gamma: 0.1
  lr: 0.007
  lr_decay_factor: 1000
  lr_scheduler: onecycle
  max_epoch: 60
  max_lr: 0.1
  min_lr: 1e-05
  mix_precision: True
  multiplier: 10
  nbd: True
  new_layers: ['classifier']
  open_layers: ['classifier']
  optim: sam
  patience: 5
  pct_start: 0.2
  print_freq: 20
  sam:
    adaptive: False
    rho: 0.05
  seed: 5
  staged_lr: False
  start_epoch: 0
  stepsize: [20]
  target_metric: test_acc
  train_patience: 5
  warmup: 5
  weight_decay: 0.0005
use_gpu: True
2023-02-01_17:29:08|
2023-02-01_17:29:08|Building train transforms ...
2023-02-01_17:29:08|+ resize to 448x448
2023-02-01_17:29:08|+ RandAugment
2023-02-01_17:29:08|+ Cutout
2023-02-01_17:29:08|+ random flip
2023-02-01_17:29:08|+ to torch tensor of range [0, 1]
2023-02-01_17:29:08|+ normalization (mean=[0.5, 0.5, 0.5], std=[0.5, 0.5, 0.5])
2023-02-01_17:29:08|Building test transforms ...
2023-02-01_17:29:08|+ resize to 448x448
2023-02-01_17:29:08|+ to torch tensor of range [0, 1]
2023-02-01_17:29:08|+ normalization (mean=[0.5, 0.5, 0.5], std=[0.5, 0.5, 0.5])
2023-02-01_17:29:08|=> Loading train dataset
2023-02-01_17:29:08|=> Loading test dataset
2023-02-01_17:29:08|
2023-02-01_17:29:08|
2023-02-01_17:29:08|  **************** Summary ****************
2023-02-01_17:29:08|  # categories      : 20
2023-02-01_17:29:08|  # train images    : 5011
2023-02-01_17:29:08|  # test images     : 5011
2023-02-01_17:29:08|  *****************************************
2023-02-01_17:29:08|
2023-02-01_17:29:08|
2023-02-01_17:29:08|Building main model: mld_efficientnetv2_s_21k
2023-02-01_17:29:09|Main model complexity: params=26,692,197 flops=22,688,426,992.0
2023-02-01_17:29:13|Successfully loaded pretrained weights from "efficientnetv2-s-oi.pth"
2023-02-01_17:29:13|** The following layers were not loaded from checkpoint: ['backbone.model.act2.weight', 'deco                                                                 der.duplicate_pooling', 'decoder.duplicate_pooling_bias', 'decoder.embed_standart.bias', 'decoder.embed_standart.                                                                 weight', 'decoder.layers.0.linear1.bias', 'decoder.layers.0.linear1.weight', 'decoder.layers.0.linear2.bias', 'de                                                                 coder.layers.0.linear2.weight', 'decoder.layers.0.multihead_attn.in_proj_bias', 'decoder.layers.0.multihead_attn.                                                                 in_proj_weight', 'decoder.layers.0.multihead_attn.out_proj.bias', 'decoder.layers.0.multihead_attn.out_proj.weigh                                                                 t', 'decoder.layers.0.norm1.bias', 'decoder.layers.0.norm1.weight', 'decoder.layers.0.norm2.bias', 'decoder.layer                                                                 s.0.norm2.weight', 'decoder.layers.0.norm3.bias', 'decoder.layers.0.norm3.weight', 'decoder.query_embed.weight']
2023-02-01_17:29:13|Building am_binary-engine
2023-02-01_17:29:13|=> Start training
/home/miniconda3/lib/python3.7/site-packages/torch/optim/lr_scheduler.py:134: UserWarning: Detected ca                                                                 ll of `lr_scheduler.step()` before `optimizer.step()`. In PyTorch 1.1.0 and later, you should call them in the op                                                                 posite order: `optimizer.step()` before `lr_scheduler.step()`.  Failure to do this will result in PyTorch skippin                                                                 g the first value of the learning rate schedule. See more details at https://pytorch.org/docs/stable/optim.html#h                                                                 ow-to-adjust-learning-rate
  "https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate", UserWarning)
2023-02-01_17:29:50|epoch: [1/60][20/59]        time 1.061 (1.859)      data 0.050 (0.249)      cls acc 0.083 (0.                                                                 072)    eta 1:49:05     main_model 12.4576 (13.7815)    loss 12.4576 (13.7815)  lr 0.000082
2023-02-01_17:30:12|epoch: [1/60][40/59]        time 1.049 (1.471)      data 0.051 (0.152)      cls acc 0.048 (0.                                                                 073)    eta 1:25:49     main_model 13.5511 (13.3524)    loss 13.5511 (13.3524)  lr 0.000122
2023-02-01_17:30:32|epoch: [1/60][59/59]        time 0.971 (1.343)      data 0.049 (0.122)      cls acc 0.036 (0.                                                                 071)    eta 1:17:55     main_model 12.1725 (12.6952)    loss 12.1725 (12.6952)  lr 0.000184
2023-02-01_17:30:33|##### Evaluating test dataset #####
2023-02-01_17:30:58|** Results (main_model) **
2023-02-01_17:30:58|mAP: 36.21%
2023-02-01_17:30:58|P_O: 41.81%
2023-02-01_17:30:58|R_O: 26.48%
2023-02-01_17:30:58|F_O: 32.42%
2023-02-01_17:30:58|mean_P_C: 41.81%
2023-02-01_17:30:58|mean_R_C: 23.38%
2023-02-01_17:30:58|mean_F_C: 28.96%
2023-02-01_17:31:23|** Results (EMA model) **
2023-02-01_17:31:23|mAP: 7.99%
2023-02-01_17:31:23|P_O: 9.14%
2023-02-01_17:31:23|R_O: 68.09%
2023-02-01_17:31:23|F_O: 16.12%
2023-02-01_17:31:23|mean_P_C: 9.15%
2023-02-01_17:31:23|mean_R_C: 68.16%
2023-02-01_17:31:23|mean_F_C: 15.78%
2023-02-01_17:31:24|Checkpoint saved to "output/voc/v4/main_model/main_model.pth.tar-1"

**Comment by sovrasov** [comment]
It's ok that trainings are finished by early stopping: we aiming to create an adaptive pipeline, so the LR schedule is always long enough to learn even a complicated dataset. Early stopping just cuts this long schedule when required.

**Comment by sovrasov** [comment]
I don't recommend usage of more that 1 gpu for EfficeintNet on such small and simple datasets like VOC.
It will cause training slowdown and possibly overfitting. Your first log with GCN looks good.

**Comment by UnityBoy** [comment]
> 



> I don't recommend usage of more that 1 gpu for EfficeintNet on such small and simple datasets like VOC. It will cause training slowdown and possibly overfitting. Your first log with GCN looks good.

okey,i will try it.thx

**Comment by sungmanc** [comment]
Close this issue, please reopen it if there is any problem.
